sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeArrayWriter.java

```diff
@@ -55,10 +55,19 @@ public void initialize(int numElements) {
 
     this.startingOffset = cursor();
 
+    long fixedPartInBytesLong =
+      ByteArrayMethods.roundNumberOfBytesToNearestWord((long) elementSize * numElements);
+    long totalInitialSize = headerInBytes + fixedPartInBytesLong;
+
+    if (totalInitialSize > Integer.MAX_VALUE) {
+      throw new IllegalArgumentException(
```

Comment:
Is it possible to trigger the error from user space (from SQL for instance). If so, please, introduce an error class and place it to error-classes.json.

Reply:
I think so, let's make the error more user-facing.
error class: TOO_MANY_ARRAY_ELEMENTS
message: "Cannot initialize array with %numElements elements"

Reply:
Just a point of clarification. The issue isn't too many elements per se. It's too many elements for the given element size.

If, in the above example, I had cast `val` as int, the `collect_list` would have succeeded (i.e., 268271216 int elements is just fine, but the same number of bigint elements is not).

How about this for a message?:
"Cannot initialize array with %numElements elements of size %size"

Caveat: my suggested message will be a little confusing for arrays of non-primitives, where the element size is always 8 (for the size/offset).